dspy/teleprompt/bootstrap_finetune.py

```diff
@@ -0,0 +1,261 @@
+        logger.info(f"Starting {num_jobs} fine-tuning jobs...")
+        # TODO(nit) Pass an identifier to the job so that we can tell the logs
+        # coming from different fine-tune threads.
+
+        key_to_job = {}
+        for key, finetune_kwargs in finetune_dict.items():
+            lm = finetune_kwargs.pop("lm")
+            key_to_job[key] = lm.finetune(**finetune_kwargs)
+        
+        key_to_lm = {}
+        for ind, (key, job) in enumerate(key_to_job.items()):
+            key_to_lm[key] = job.result()
+            job.thread.join()
+            logger.info(f"Job {ind + 1}/{num_jobs} completed.")
+
+        return key_to_lm
+
+    def _prepare_finetune_data(self, trace_data: List[Dict[str, Any]], lm: LM, pred_ind: Optional[int] = None):
+        # TODO(nit) Log dataset details/size
+        if self.metric:
+            trace_data = [d for d in trace_data if d["score"]]
+
+        data = []
+        adapter = self.adapter[lm] or lm.infer_adapter()
+        data_format = infer_data_format(adapter)
+        for item in trace_data:
+            for pred_ind, _ in enumerate(item['trace']):
+                include_data = pred_ind == None or pred_ind == pred_ind
+                if include_data:
+                    call_data = build_call_data_from_trace(trace=item['trace'], pred_ind=pred_ind,  adapter=adapter, exclude_demos=self.exclude_demos)
+                    data.append(call_data)
+
+        return data, data_format
+
+
+def build_call_data_from_trace(
+    trace: List[Dict],
+    pred_ind: int,
+    adapter: Optional[Adapter] = None,
+    exclude_demos: bool = False,
+) -> Dict[str, List[Dict[str, Any]]]:
+    # Find data that's relevant to the predictor
+    pred, inputs, outputs = trace[pred_ind]  # assuming that the order is kept
+
+    if not adapter:
+        # TODO(feature): A trace is collected using a particular adapter. It
+        # would be nice to get this adapter information from the trace (e.g.
+        # pred.lm.adapter) as opposed to using the inference method below.
+        adapter = pred.lm.infer_adapter()
+
+    demos = [] if exclude_demos else pred.demos
+    call_data = adapter.format_finetune_data(
+        signature=pred.signature,
+        demos=demos,
+        inputs=inputs,
+        outputs=outputs,
+    )
+    return call_data
+
+
+def bootstrap_trace_data(
+    program: Program,
+    dataset: List[Example],
+    metric: Optional[Callable] = None,
+    num_threads=6,
+) -> List[Dict[str, Any]]:
+    # Return a list of dicts with the following keys:
+    #     example_ind, example, prediction, trace, and score (if metric != None)
+    evaluator = Evaluate(
+        devset=dataset, num_threads=num_threads, display_progress=True,
+        provide_traceback=True  # TODO(check with team)
+    )
+    # TODO(PR): Should "trace" not be included in the lambda function?
+    _metric = metric if metric else lambda example, prediction: 1
+    evaluator(program, metric=_metric)
```

Comment:
Why are doing a full eval here?

Reply:
I was using the evaluator to make use of the threaded sampling